protocols/gossipsub/src/config.rs

```diff
@@ -302,33 +669,200 @@ impl GossipsubConfigBuilder {
+        self.config.gossip_retransimission = gossip_retransimission;
+        self
+    }
+
+    /// The maximum number of new peers to graft to during opportunistic grafting. The default is 2.
+    pub fn opportunistic_graft_peers(&mut self, opportunistic_graft_peers: usize) -> &mut Self {
+        self.config.opportunistic_graft_peers = opportunistic_graft_peers;
+        self
+    }
+
+    /// The maximum number of messages to include in an IHAVE message.
+    /// Also controls the maximum number of IHAVE ids we will accept and request with IWANT from a
+    /// peer within a heartbeat, to protect from IHAVE floods. You should adjust this value from the
+    /// default if your system is pushing more than 5000 messages in GossipSubHistoryGossip
+    /// heartbeats; with the defaults this is 1666 messages/s. The default is 5000.
+    pub fn max_ihave_length(&mut self, max_ihave_length: usize) -> &mut Self {
+        self.config.max_ihave_length = max_ihave_length;
+        self
+    }
+
+    /// GossipSubMaxIHaveMessages is the maximum number of IHAVE messages to accept from a peer
+    /// within a heartbeat.
+    pub fn max_ihave_messages(&mut self, max_ihave_messages: usize) -> &mut Self {
+        self.config.max_ihave_messages = max_ihave_messages;
+        self
+    }
+
+    pub fn iwant_followup_time(&mut self, iwant_followup_time: Duration) -> &mut Self {
+        self.config.iwant_followup_time = iwant_followup_time;
+        self
+    }
+
+    /// Enable support for flooodsub peers.
+    pub fn support_floodsub(&mut self) -> &mut Self {
+        self.config.support_floodsub = true;
+        self
+    }
+
+    /// Constructs a `GossipsubConfig` from the given configuration and validates the settings.
+    pub fn build(&self) -> Result<GossipsubConfig, &str> {
+        // check all constraints on config
+
+        if self.config.max_transmit_size < 100 {
+            return Err("The maximum transmission size must be greater than 100 to permit basic control messages");
+        }
+
+        if !(self.config.history_length >= self.config.history_gossip) {
+            return Err(
+                "The history_length must be greater than or equal to the history_gossip \
+                length",
+            );
+        }
+
+        if !(self.config.mesh_outbound_min < self.config.mesh_n_low
+            && self.config.mesh_n_low <= self.config.mesh_n
+            && self.config.mesh_n <= self.config.mesh_n_high)
+        {
+            return Err("The following inequality doesn't hold \
+                mesh_outbound_min < mesh_n_low <= mesh_n <= mesh_n_high");
+        }
+
+        if !(self.config.mesh_outbound_min * 2 <= self.config.mesh_n) {
+            return Err(
+                "The following inequality doesn't hold mesh_outbound_min <= self.config.mesh_n / 2",
+            );
+        }
+        Ok(self.config.clone())
     }
 }
 
 impl std::fmt::Debug for GossipsubConfig {
```

Comment:
As far as I can tell none of the fields are treated special. Why not derive `Debug`?

Reply:
I think the derive cannot handle the `message_id_fn` and `fast_message_id_fn` types in general. So we exclude them in a custom implementation.